mujoco: 2.3.7 -> 3.0.1, add Python bindings

tmplt opened the pull request:
## Description of changes
- Bumps `mujoco` to [3.0.1](https://mujoco.readthedocs.io/en/stable/changelog.html)
- Introduces the native Python bindings, [available on PyPi](https://pypi.org/project/mujoco/3.0.1/).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

So far I've tested `bin/simulate` and importing mujoco in Python. I'm gonna play around with it a bit before I sign off on it.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tmplt commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,230 @@
-    hash = "sha256-YGAe4+Ttv/xeou+9FoJjmQCKgzupTYdDhd+gzvtz/88=";
-  };
+{ cmake, fetchFromGitHub, fetchFromGitLab, git, libGL, stdenv, xorg, fetchpatch
+, SDL2, glm, spdlog, cereal_1_3_2, eigen, glfw-lib, glfw, lib, fetchPypi
```
`glfw-lib` is here the regular system library and `glfw` is the Python module. Is there a less confusing notation?

tmplt commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,230 @@
-    rev = version;
-    hash = "sha256-LgpA+iPGqciHuWBSD6/7yvZ7p+vo48ZYKjjrDZSnAwE=";
+      nativeBuildInputs = [ cmake ];
+      buildInputs = [ glm spdlog cereal_1_3_2 pin.tmd ];
```
`glm` is not pinned here because the 2018 release that sdflib uses does not build due to missing files and test failures.

tmplt replied:
sdflib does not build with `cereal 1.3.0` because a `cereal::cereal` link target cannot be found. 1.3.2 is also the version that is pinned upstream.

tmplt commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,230 @@
+          concatStrings (reverseList
+            (drop 1 (reverseList (splitString "." python3.version))));
+      in ''
+        sed -i '140d;141d;142d;143d;173d;174d;175d;176d' mujoco/CMakeLists.txt # abseil, eigen
```
This removes
```
  GIT_REPO
  https://github.com/abseil/abseil-cpp
  GIT_TAG
  fb3621f4f897824c0dbe0615fa94543df6192f30 # LTS 20230802.1
```
and similar for abseil and eigen.

There is probably a cleaner call.

tmplt commented on pkgs/top-level/all-packages.nix:
```diff
@@ -24082,7 +24082,14 @@ with pkgs;
+    inherit (python3Packages) buildPythonPackage setuptools setuptools-scm wheel pybind11;
+    glfw-lib = glfw;
+    glfw = python3Packages.glfw;
+    }).mujoco;
```
How can I simplify this? An argument-less `callPackage` did not work as expected.

tmplt commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7119,6 +7119,15 @@ self: super: with self; {
+      inherit (self)  buildPythonPackage setuptools setuptools-scm wheel pybind11;
+      glfw-lib = glfw;
+      glfw = self.glfw;
+    }).python-bindings;
```
ditto.

tmplt commented on pkgs/development/python-modules/glfw/default.nix:
Whoops. Forgot I fixed this build.

tmplt replied:
Rebased.

samuela commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -12,35 +7,29 @@ buildPythonPackage rec {
-    repo = "pyGLFW";
-    rev = "refs/tags/v${version}";
-    hash = "sha256-3K+mDSz4ifVYkUvhd2XDPbhh6UCY4y54YqNLoAYDsP0=";
+  src = fetchPypi {
```
why `fetchPypi` instead of `fetchFromGitHub`? building from source is always preferred

tmplt replied:
I was under the impression it was the other way around. Reverted.

samuela commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -12,35 +7,29 @@ buildPythonPackage rec {
     description = "Python bindings for GLFW";
     homepage = "https://github.com/FlorianRhiem/pyGLFW";
-    changelog = "https://github.com/FlorianRhiem/pyGLFW/blob/v${version}/CHANGELOG.md";
+    changelog =
+      "https://github.com/FlorianRhiem/pyGLFW/blob/v${version}/CHANGELOG.md";
```
nit: unnecessary, let's keep the git blame clean

samuela commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -12,35 +7,29 @@ buildPythonPackage rec {
-  pythonImportsCheck = [
-    "glfw"
-  ];
+  pythonImportsCheck = [ "glfw" ];
```
nit: unnecessary, let's keep the git blame clean

samuela commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -12,35 +7,29 @@ buildPythonPackage rec {
-  propagatedBuildInputs = [
-    glfw3
-  ];
+  propagatedBuildInputs = [ glfw3 ];
```
nit: unnecessary, let's keep the git blame clean

samuela commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -1,9 +1,4 @@
-, glfw3
-, pythonOlder
-}:
+{ lib, buildPythonPackage, fetchFromGitHub, fetchPypi, glfw3, pythonOlder }:
```
nit: unnecessary, let's keep the git blame clean

samuela commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,236 @@
-    rev = "9a89766acc42ddfa9e7133c7d81a5bda108a0ade";
-    hash = "sha256-YGAe4+Ttv/xeou+9FoJjmQCKgzupTYdDhd+gzvtz/88=";
-  };
+{ cmake, fetchFromGitHub, fetchFromGitLab, libGL, stdenv, xorg, fetchpatch, SDL2
+, glm, spdlog, cereal_1_3_2, eigen, glfw, lib, fetchPypi, python3Packages
+, python3 }:
```
nit: unnecessary, let's keep the git blame clean

samuela commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,236 @@
+    inherit pname version;
+    format = "pyproject";
+
+    src = fetchPypi {
```
why not `fetchFromGitHub`?

tmplt replied:
Because the PyPi tarball is impurely built via [`python/make_sdist.sh`](https://github.com/google-deepmind/mujoco/blob/main/python/make_sdist.sh) in the project's CI. I suppose it can be patched, but it was nothing I wanted to handle this first pass.

samuela replied:
Ah ok thanks for clarifying! in that case, let's leave a code comment explaining this situation

samuela commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,236 @@
+
+  python-bindings = python3Packages.buildPythonPackage rec {
+    inherit pname version;
+    format = "pyproject";
```
```suggestion
    pyproject = true;
```

samuela reviewed (changes requested):
Thanks for putting this together @tmplt ! Let's split the python module out into its own derivation in `python-modules/`. That should simplify things substantially, and keeps things in line with nixpkgs convention. `passthru` can be used to help usher things like `pin.eigen3`, etc across the package boundary.

tmplt committed:
glfw: fix build for Python bindings

tmplt committed:
mujoco: 2.3.7 -> 3.0.0, add Python bindings

https://mujoco.readthedocs.io/en/stable/changelog.html#version-3-0-0-october-18-2023

Fixes #262698
Fixes #262629
Fixes #262593

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

tmplt commented:
Wasn't aware of `passthru`. Should now follow convention.

tmplt commented:
Currently missing is MJX support (e.g. `from mujoco import mjx`) for GPU acceleration.

samuela commented on pkgs/development/python-modules/mujoco/default.nix:
```diff
@@ -0,0 +1,76 @@
+, python3
+}:
+
+let mujoco = pkgs.mujoco;
```
```suggestion
```
let's add `mujoco` to the arguments list instead

samuela commented on pkgs/development/python-modules/mujoco/default.nix:
```diff
@@ -0,0 +1,76 @@
+}:
+
+let mujoco = pkgs.mujoco;
+in python3Packages.buildPythonPackage rec {
```
```suggestion
in buildPythonPackage rec {
```

samuela commented on pkgs/development/python-modules/mujoco/default.nix:
```diff
@@ -0,0 +1,76 @@
+    hash = "sha256-2ImKDfm9hCORaH7iou/Lp979VSCugCM/Gqnb/CkTeOw=";
+  };
+
+  nativeBuildInputs = with python3Packages; [ cmake setuptools ];
```
```suggestion
  nativeBuildInputs = [ cmake setuptools ];
```

samuela commented on pkgs/development/python-modules/mujoco/default.nix:
```diff
@@ -0,0 +1,76 @@
+    # dependency directory. $build from setuptools.
+    (let
+      # E.g. 3.11.2 -> "311"
+      pythonVersionMajorMinor = with lib.strings;
+        with lib.lists;
+        concatStrings
+        (reverseList (drop 1 (reverseList (splitString "." python3.version))));
```
```suggestion
      pythonVersionMajorMinor = with lib.versions; "${major pythonVersion}${minor pythonVersion}";
```
something like this ought to do the trick. check out [lib.versions](https://github.com/NixOS/nixpkgs/blob/master/lib/versions.nix) for other helpy helpers

samuela reviewed (changes requested):
left a bunch of comments, but they're each pretty straightforward and just focused on adhering to nixpkgs conventions. thanks for all the hard work on this!

tmplt commented on pkgs/development/python-modules/mujoco/default.nix:
```diff
@@ -0,0 +1,76 @@
+  preConfigure =
+    # Use system packages for pybind, glfw
+    ''
+      sed -i '185s/.*/  ON/' mujoco/CMakeLists.txt # pybind
+      sed -i '79s/.*/  ON/'  mujoco/simulate/cmake/SimulateDependencies.cmake # glfw
```
I'm bad at multi-line regex and we're flipping only two `OFF`s to `ON`s.

tmplt replied:
Learned some regex.

tmplt commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,179 @@
 
   buildInputs = [
-    libGL
-    xorg.libX11
-    xorg.libXcursor
-    xorg.libXext
-    xorg.libXi
-    xorg.libXinerama
-    xorg.libXrandr
```
No longer necessary, at least for `bin/simulate`. Everything seems to go through GLFW.

tmplt commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,179 @@
-    rev = "9a89766acc42ddfa9e7133c7d81a5bda108a0ade";
-    hash = "sha256-YGAe4+Ttv/xeou+9FoJjmQCKgzupTYdDhd+gzvtz/88=";
-  };
+  pin = {
```
It made it easier while writing it to check of which deps I had yet to pin, but I suppose it can be removed now.

tmplt replied:
I'd prefer to keep it explicit which dependencies are pin-matched with upstream.

samuela replied:
ok fair enough

tmplt commented on pkgs/development/python-modules/glfw/default.nix:
```diff
@@ -20,9 +20,8 @@ buildPythonPackage rec {
-  patches = [ ./search-path.patch ];
-
   postPatch = ''
+    sed -i "255s/_get_library_search_paths()/\['@GLFW@'\]/" glfw/library.py
```
Any suggestions? I specify line number because I don't want to replace the function definition.

samuela replied:
you could either add a patch file or do eg
```
substituteInPlace glfw/library.py --replace "_get_library_search_paths()," "[ '${glfw3}/lib' ]"
```

samuela commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7119,6 +7119,11 @@ self: super: with self; {
 
+  mujoco = callPackage ../development/python-modules/mujoco {
+    inherit (pkgs) mujoco;
+    pythonVersion = python.version;
```
Ah I led you slightly astray. I was under the impression that `pythonVersion` was automatically passed as an argument to python module derivations, but it lives as `python.pythonVersion`. see eg [here](https://github.com/NixOS/nixpkgs/blob/67c0a676dd5e9998beb12d4f79ec44d676181802/pkgs/development/python-modules/torchvision/bin.nix#L16)

samuela commented on pkgs/applications/science/robotics/mujoco/default.nix:
```diff
@@ -1,122 +1,184 @@
+    ln -s ${pin.marchingcubecpp} build/_deps/marchingcubecpp-src
   '';
 
+  passthru.pin = { inherit (pin) lodepng eigen3 abseil-cpp; };
```
```suggestion
  passthru = { inherit pin; };
```
any reason not to just pass all of `pin`?

tmplt replied:
Just to keep it explicit. Only a subset of the pins are required for the Python bindings.

samuela replied:
ok fair enough

samuela reviewed (approved):
diff LGTM!

tmplt commented:
[3.0.1](https://mujoco.readthedocs.io/en/3.0.1/changelog.html) is now out.

tmplt committed:
mujoco: 3.0.0 -> 3.0.1

samuela commented:
I think this is ready to go --- shall we mark as "ready for review"?

samuela commented:
Thanks so much @tmplt !

tmplt commented:
For future consideration: https://discourse.nixos.org/t/dealing-with-cmake-fetchcontent/34768
